File: src/main/kotlin/dev/jefy/connectpro/user/domain/model/Token.java
package dev.jefy.connectpro.user.domain.model;

import org.springframework.util.Assert;

import java.time.Instant;

import dev.jefy.connectpro.user.domain.vo.OtpCode;
import dev.jefy.connectpro.user.domain.vo.TokenId;
import dev.jefy.connectpro.user.domain.vo.UserId;
import jakarta.persistence.*;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.NoArgsConstructor;

/**
 * @author Jôph Yamba
 */
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@Entity
@Table(name = "tokens")
public class Token {
    @EmbeddedId
    @AttributeOverride(name = "value", column = @Column(name = "id"))
    private TokenId id;
    
    @Embedded
    @AttributeOverride(name = "value", column = @Column(name = "otp_code"))
    private OtpCode code;
    
    private Instant createdAt;
    
    private Instant expireAt;
    
    private Instant validatedAt;
    
    @Embedded
    @AttributeOverride(name = "value", column = @Column(name = "user_id"))
    private UserId userId;

    public Token(UserId userId, OtpCode code, long verificationTokenExpirationTimeInMinutes) {
        Assert.notNull(userId, "reviewerId cannot be null");
        Instant createdDate = Instant.now();
        Instant expiresAt = createdDate.plusSeconds(verificationTokenExpirationTimeInMinutes * 60);
        this.id = TokenId.generate();
        this.code = code;
        this.createdAt = createdDate;
        this.expireAt = expiresAt;
        this.validatedAt = null;
        this.userId = userId;
    }
    
    public boolean isExpired() {
        return Instant.now().isAfter(expireAt);
    }

    public void validate() {
        this.validatedAt = Instant.now();
    }

    public boolean isNotValidated() {
        return validatedAt == null;
    }
}

File: src/main/kotlin/dev/jefy/connectpro/user/domain/model/Token.kt
package dev.jefy.connectpro.user.domain.model

import dev.jefy.connectpro.user.domain.vo.OtpCode
import dev.jefy.connectpro.user.domain.vo.TokenId
import dev.jefy.connectpro.user.domain.vo.UserId
import jakarta.persistence.*
import java.time.Instant

@Entity
@Table(name = "tokens")
open class Token(userId: UserId, code: OtpCode, otpTokenExpiration: Long) {

    @EmbeddedId
    @AttributeOverride(name = "value", column = Column(name = "id"))
    var id: TokenId = TokenId.generate()
        protected set

    @Embedded
    @AttributeOverride(name = "value", column = Column(name = "otp_code"))
    var code: OtpCode = code
        protected set

    var createdAt: Instant = Instant.now()
        protected set 
    
    var expireAt: Instant = createdAt.plusSeconds(otpTokenExpiration * 60)
        protected set

    var validatedAt: Instant? = null
        protected set

    @Embedded
    @AttributeOverride(name = "value", column = Column(name = "user_id"))
    var userId: UserId = userId
        protected set
    
    fun isExpired(): Boolean = Instant.now().isAfter(expireAt)

    fun validate() { validatedAt = Instant.now() }

    fun isNotValidated(): Boolean = validatedAt == null
}